### PR TITLE
Dann

### PR DESCRIPTION
**Pull request by olliethomas**

### CODE REVIEW

File: bolts/models/dann.py [hunk 1]
```diff
@@ -100,18 +100,18 @@ def _inference_epoch_end(
...
     def _inference_step(self, batch: DataBatch, stage: Stage) -> Dict[str, Tensor]:
         _s, _y = self.forward(batch.x)
-        target_s = batch.s.unsqueeze(-1).float() if len(batch.s.shape) == 1 else batch.y.float()
+        target_s = batch.s.view(-1, 1).float()
         loss_adv = self._loss_adv_fn(_s, target_s)
-        target_y = batch.y.unsqueeze(-1).float() if len(batch.y.shape) == 1 else batch.y.float()
+        target_y = batch.y.view(-1, 1).float()
         loss_clf = self._loss_clf_fn(_y, target_y)
         loss = loss_adv + loss_clf
         tm_acc_s = self.val_acc_s if stage == "val" else self.test_acc_s
```
Comment (MylesBartlett) [inline bolts/models/dann.py:108]: how about using `getattr(f"{stage}_acc_s")` here?

File: bolts/models/dann.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    @staticmethod
+    def forward(ctx: autograd.Function, x: Tensor, lambda_: float) -> Tensor:
+        """Do GRL."""
+        ctx.lambda_ = lambda_
+        return x.view_as(x)
+
+    @staticmethod
+    def backward(ctx: autograd.Function, grad_output: Tensor) -> Tuple[Tensor, Optional[Tensor]]:
+        """Do GRL."""
+        return grad_output.neg().mul(ctx.lambda_), None
```
Comment (MylesBartlett) [inline bolts/models/dann.py:32]:
```suggestion
        return -ctx.lambda_ * grad_output, None
```

File: bolts/models/dann.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+
+class GradReverse(autograd.Function):
+    """Gradient reversal layer."""
+
+    @staticmethod
+    def forward(ctx: autograd.Function, x: Tensor, lambda_: float) -> Tensor:
+        """Do GRL."""
+        ctx.lambda_ = lambda_
+        return x.view_as(x)
```
Comment (MylesBartlett) [inline bolts/models/dann.py:27]: why?

File: bolts/models/dann.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+__all__ = ["Dann"]
+
+Stage = Literal["train", "val", "test"]
+
+
+class GradReverse(autograd.Function):
+    """Gradient reversal layer."""
+
+    @staticmethod
+    def forward(ctx: autograd.Function, x: Tensor, lambda_: float) -> Tensor:
```
Comment (MylesBartlett) [inline bolts/models/dann.py:24]:
```suggestion
    def forward(ctx: autograd.Function, x: Tensor, lambda_: float = 1.0) -> Tensor:
```

File: bolts/models/dann.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+        self._loss_adv_fn = nn.BCEWithLogitsLoss()
+        self._loss_clf_fn = nn.BCEWithLogitsLoss()
+
+        self.test_acc_s = torchmetrics.Accuracy()
+        self.test_acc_y = torchmetrics.Accuracy()
+        self.train_acc_s = torchmetrics.Accuracy()
+        self.train_acc_y = torchmetrics.Accuracy()
+        self.val_acc_s = torchmetrics.Accuracy()
+        self.val_acc_y = torchmetrics.Accuracy()
```
Comment (MylesBartlett) [inline bolts/models/dann.py:69]: Maybe store these metric-trackers in a dictionary?

File: bolts/models/dann.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                f"{stage}/loss_clf": loss_clf.item(),
+                f"{stage}/acc_s": s_acc,
+                f"{stage}/acc_y": y_acc,
+            }
+        )
+        return {"y": batch.y, "s": batch.s, "preds": _y.sigmoid().round().squeeze(-1)}
+
+    def reset_parameters(self) -> None:
+        """Reset the models."""
+        self.adv.apply(self._maybe_reset_parameters)
```
Comment (MylesBartlett) [inline bolts/models/dann.py:129]: apply is recursive - it's sufficient just to 'apply' `_maybe_reset_parameters` to self

File: bolts/models/dann.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    def training_step(self, batch: DataBatch, batch_idx: int) -> Tensor:
+        _s, _y = self.forward(batch.x)
+        target_s = batch.s.view(-1, 1).float()
+        loss_adv = self._loss_adv_fn(_s, target_s)
+        target_y = batch.y.view(-1, 1).float()
+        loss_clf = self._loss_clf_fn(_y, target_y)
+        loss = loss_adv + loss_clf
+
+        target_s = batch.s.view(-1, 1).long()
+        acc_s = self.train_acc_s(_s >= 0, target_s)
```
Comment (MylesBartlett) [inline bolts/models/dann.py:162]: I think there should probably be a wrapper around Accuracy to save us calling >= 0 all the time

File: bolts/models/dann.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+        super().__init__()
+        self.grl_lambda = grl_lambda
+        self.learning_rate = lr
+        self.weight_decay = weight_decay
+
+        self.adv = adv
+        self.enc = enc
+        self.clf = clf
+
+        self._loss_adv_fn = nn.BCEWithLogitsLoss()
```
Comment (MylesBartlett) [inline bolts/models/dann.py:61]: Could condense to a single _loss_fn attribute

File: bolts/models/dann.py (hunk 3 above)
Comment (olliethomas) [reply bolts/models/dann.py:27]: you know what - I don't know

**Review by MylesBartlett — APPROVED** [review]
